File: training/train on 2100-2200 problems/E_Array_Shrinking.cpp
#include <bits/stdc++.h>
#define ll long long
using namespace std;

vector<int> a;
map<vector<int>,int> dp;

int solve(vector<int> a){
    if (dp.count(a)){
        return dp[a];
    }
    int s=0;
    for (int i=0;i<a.size()-1;i++){
        if (a[i]==a[i+1]){
            s=1; 
            break;
        }
    }
    if (s==0) return dp[a]=a.size();
    int res=1e9;
    for (int i=0;i<a.size()-1;i++){
        if(a[i]==a[i+1]){
            int e=a[i+1];
            a[i]++;
            a.erase(a.begin()+i+1);
            res=min({solve(a),res});
            a[i]--;
            a.insert(a.begin()+i+1,e);
        }
        
    }
    return dp[a]=res;
}

int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    int n;
    cin>>n;
    for (int i=0;i<n;i++){
        int v;
        cin>>v;
        a.push_back(v);
    }
    int res=solve(a);
    cout<<res<<"\n";
}

File: training/train on 2100-2200 problems/E_Array_Shrinking.py
def solve(indx):
    if indx>=len(a)-1:
        return len(a)
    if dp[indx]!=-1:
        return dp[indx]
    ch1=float("inf")
    ch2=float("inf")
    ch3=float("inf")
    if a[indx]==a[indx+1]:
        i=a.pop(indx+1)
        a[indx]+=1
        ch1=solve(indx)
        a[indx]-=1
        a.insert(indx+1,i)
    if indx>0 and a[indx-1]==a[indx]:
        i=a.pop(indx-1)
        ch2=solve(indx-1)
        a.insert(indx-1,i)
    ch3=solve(indx+1)
    res=min(ch1,ch2,ch3)
    dp[indx]=res
    return res


n=int(input())
a=list(map(int,input().split()))
dp=[-1 for _ in range(n)]
res=solve(0)
print(res)